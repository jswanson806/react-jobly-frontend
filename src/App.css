body {
  font-family: 'Raleway', sans-serif;
}

main {
  width: 100vw;
  display: flex;
  flex-direction: column;
  align-items: center;
  padding-top: 15vh;
  height: calc(100vh - 56px); /* router view - navbar height */
  overflow: scroll;
  color: #000000; /* black */

  /* apply colors to background */
  background-color: #c5a880; /* Khaki */
  background-image: linear-gradient(135deg, #c5a880 0%, #e6d8af 100%);
}

.card {
  border-radius: 0;
  box-shadow: 0 2px 2px 0 rgba(0, 0, 0, 0.14),
    0 3px 1px -2px rgba(0, 0, 0, 0.12), 0 1px 5px 0 rgba(0, 0, 0, 0.2);
  margin-bottom: 2em;

  /* apply color to border */
  border: 2px solid #8c4646; /* Firebrick */
}
