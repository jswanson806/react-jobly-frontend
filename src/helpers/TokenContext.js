import React, { createContext, useState } from 'react';

// Create a new context for the token
export const TokenContext = createContext();

// Define the initial state for the token
const INITIAL_TOKEN_STATE = '';

// Create a provider component for the token context
export const TokenProvider = ({ children }) => {
  // Set up state to store the token value
  const [token, setToken] = useState(INITIAL_TOKEN_STATE);

  // Render the provider component and pass the token value and setter to the context
  return (
    <TokenContext.Provider value={{ token, setToken }}>
      {children}
    </TokenContext.Provider>
  );
};
