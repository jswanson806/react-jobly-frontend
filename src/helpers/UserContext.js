// Import the necessary modules from React
import React, { createContext, useState } from 'react';

// Define the initial state for the user
const INITIAL_USER_STATE = {};

// Create a new context for the user
export const UserContext = createContext();

// Create a provider component for the user context
export const UserProvider = ({ children }) => {
  // Set up state to store the user object
  const [user, setUser] = useState(INITIAL_USER_STATE);

  // Render the provider component and pass the user object and setter to the context
  return (
    <UserContext.Provider value={{ user, setUser }}>
      {children}
    </UserContext.Provider>
  );
};

