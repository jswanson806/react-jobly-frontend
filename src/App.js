import React, { useContext, useEffect } from 'react';
import { useHistory } from "react-router-dom";
import './App.css';
import Routes from './Routes.js';
import NavBar from './NavBar.js';
import JoblyApi from './api.js';
import AppContext from './helpers/AppContext.js';
import { UserContext } from './helpers/UserContext.js';
import { TokenContext } from './helpers/TokenContext.js';

/**
 * App Component
 *
 * The root component of the application. Manages user authentication, routing, and global state.
 *
*/

function App() {

  const history = useHistory();

  // ********************* State *********************

  const {user, setUser} = useContext(UserContext);
  const {token, setToken} = useContext(TokenContext);

  // ********************* Hooks *********************

  // keep token saved between page refreshes
  // set token to the value in localStorage
  useEffect(() => {
    // get the value of 'token' from localStorage
    const storedToken = localStorage.getItem('token');
    const storedUser = localStorage.getItem('user');

    // check for value of stored token to be different than stateful token
    if(storedToken !== token && storedToken !== null) {
      // set token to the value in localStorage
      setToken(storedToken);
    }
  // check for value of stored user to be different than stateful user
    if(storedUser !== user && storedUser !== null) {
      // turn the stringified user object in localStorage into an object and set user state
      setUser(JSON.parse(storedUser))
    }

    // only run on first render
  }, [])

  // ********************* Functions *********************

  // registers a new user and logs them in
  const registerUser = async (userInfo) => {
    // api method registers user
    await JoblyApi.registerUser(userInfo);
    // calls logIn function to log in the newly registered user
    logIn(userInfo);
  }

  // logs in an existing user
  const logIn = async ({username, password}) => {
    // authenticates user
    let resToken = await JoblyApi.logInUser(username, password);
    // set token in localStorage and state
    localStorage.setItem('token', resToken);
    setToken(resToken);
    // setLocalToken(resToken)
    // sets api token to token received after authentication
    JoblyApi.token = resToken;
    // get the user from the database
    let userRes = await JoblyApi.getUser(username);
    // stringify user object and set user info in localStorage and state
    localStorage.setItem('user', JSON.stringify(userRes.user));
    setUser(userRes.user);
  }

  // logs out user
  const logOut = async () => {
    // reset api token
    JoblyApi.token = "";
    // clear localStorage'token' and 'user'
    localStorage.removeItem('token');
    localStorage.removeItem('user');
    // reset state of 'token' and 'user'
    setToken("")
    setUser("")
    // redirect to home
    history.push('/');
  }

  // calls api to patch user info
  const updateUserProfile = async (newUserInfo) => {
    const {applications} = user;
    await JoblyApi.patchUser(newUserInfo);
    // set user info state and user info in localStorage to new user info
    // res.user.applications = user.applications
    newUserInfo.applications = applications;
    setUser(newUserInfo);
  }

  // ********************* Return *********************

  return (
    // AppContext provides functions and user and token state to child components
    <AppContext.Provider value={{user, setUser, token, setToken, registerUser, logIn, logOut, updateUserProfile}} >
          <div data-testid="app">
            <NavBar />
            <Routes />
          </div>
    </AppContext.Provider>
  )
}

export default App;