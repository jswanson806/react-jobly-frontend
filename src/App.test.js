import React from 'react';
import { act, render, screen, waitFor } from '@testing-library/react';
import '@testing-library/jest-dom';
import App from './App.js';
import { UserContext } from './helpers/UserContext.js';
import { TokenContext } from './helpers/TokenContext.js';



// ************************* Setup *************************

let user = {
  username: "testuser",
  password: "password",
  firstName: "firstName",
  lastName: "lastName",
  email: "test@email.com",
  applications: []
}

jest.mock('./Routes.js', () => {
  return () => <div data-testid="mocked-routes" />;
});

jest.mock('./NavBar.js', () => {
  return () => <div data-testid="mocked-navbar"/>
});

// ************************* Tests *************************

describe('App component smoke and snapshot', () => {

  it('should render the App component', async () => {
    // ensure that all state updates and side effects have been processed by wrapping in act()
    await act(async () => {render(
      // wrapping in MemoryRouter to provide a context for Route to match URL
        <UserContext.Provider value={{user}}>
          <TokenContext.Provider value={""}>
            <App />
          </TokenContext.Provider>
        </UserContext.Provider>
      );
    });
   
  });

  it('should match the snapshot', async () => {
    let asFragment;

    await act(async () => {const {asFragment: fragment} = render(
      <UserContext.Provider value={{user}}>
      <TokenContext.Provider value={""}>
          <App />
        </TokenContext.Provider>
      </UserContext.Provider>
      );
      asFragment = fragment;
    });

    expect(asFragment()).toMatchSnapshot();

  });

});


describe('Child component rendering', () => {

  it('should render the NavBar component', async () => {
    await act( async () => {
      render(
        <UserContext.Provider value={{user}}>
          <TokenContext.Provider value={""}>
              <App />
          </TokenContext.Provider>
        </UserContext.Provider>
      );
    });

    const navbarComponent = screen.getByTestId("mocked-navbar");
    expect(navbarComponent).toBeInTheDocument();

  });

  it('should render the Routes component', async () => {
    await act( async () => {
      render(
        <UserContext.Provider value={{user}}>
          <TokenContext.Provider value={""}>
            <App />
          </TokenContext.Provider>
        </UserContext.Provider>
      )  
    });

      const navbarComponent = screen.getByTestId('mocked-routes');
      expect(navbarComponent).toBeInTheDocument();
    
  });


});
