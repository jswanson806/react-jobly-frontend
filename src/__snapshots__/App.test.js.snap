// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App component smoke and snapshot should match the snapshot 1`] = `
<DocumentFragment>
  <div
    data-testid="app"
  >
    <div
      data-testid="mocked-navbar"
    />
    <div
      data-testid="mocked-routes"
    />
  </div>
</DocumentFragment>
`;
